mipengine/node/monetdb_interface/monet_db_connection.py

```diff
@@ -0,0 +1,92 @@
+    """
+    Copied from https://stackoverflow.com/questions/6760685/creating-a-singleton-in-python
+    """
+
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class MonetDB(metaclass=Singleton):
+    """
+    MonetDB is a Singleton class because we want it to be initialized at runtime.
+
+    If the connection is a public module variable, it will be initialized at import time
+    from Celery and all the Celery workers will use the same connection instance.
+
+    We want one MonetDB connection instance per Celery worker/process.
+
+    """
+
+    def __init__(self):
+        global_node = node_catalog.get_global_node()
+        if global_node.nodeId == config.get("node", "identifier"):
+            node = global_node
+        else:
+            node = node_catalog.get_local_node_data(config.get("node", "identifier"))
+        monetdb_hostname = node.monetdbHostname
+        monetdb_port = node.monetdbPort
+        self._connection = pymonetdb.connect(
+            username=config.get("monet_db", "username"),
+            port=monetdb_port,
+            password=config.get("monet_db", "password"),
+            hostname=monetdb_hostname,
+            database=config.get("monet_db", "database"),
+        )
+
+    def _get_connection(self):
+        """
+        Commits the connection and then retrieves it so it is up-to-date.
+        """
+        self._connection.commit()
+        return self._connection
+
+    def execute_with_result(self, query: str) -> List:
+        """
+        Used to execute select queries that return a result.
+
+        Should NOT be used to execute "CREATE, DROP, ALTER, UPDATE, ..." statements.
+        """
+        cursor = self._get_connection().cursor()
+        cursor.execute(query)
+        result = cursor.fetchall()
+        cursor.close()
+        return result
+
+    def execute(self, query: str):
+        """
+        Executes statements that don't have a result. For example "CREATE,DROP,UPDATE".
+        And handles the *Optimistic Concurrency Control by giving each call X attempts
+        if they fail with pymonetdb.exceptions.IntegrityError .
+        *https://www.monetdb.org/blog/optimistic-concurrency-control
+        """
+        attempts = 0
+        connection = self._get_connection()
+        while attempts <= OCC_MAX_ATTEMPTS:
+            cursor = connection.cursor()
+            try:
+                cursor.execute(query)
+                connection.commit()
+                break
+            except pymonetdb.exceptions.IntegrityError as integrity_exc:
+                connection.rollback()
+                attempts += 1
+                if attempts >= OCC_MAX_ATTEMPTS:
+                    raise integrity_exc
+            except Exception as exc:
+                connection.rollback()
+                raise exc
```

Comment:
Do you have a specific use case in mind? To be honest I am not sure how `for/else` could be used here.

We don't want something to be executed when the loop finishes.

We want to propagate the integrity exception only if this is the last iteration.

Reply:
I agree that it looks prettier but not that someone reading it will understand it faster... @KFilippopolitis it's up to you I guess...

@jassak with this code, how will you propagate the same exception that you received?
```
else: 
    raise integrity_exc
```
will not work since integrity_exc is not assigned. Maybe we can throw a `pymonetdb.exceptions.IntegrityError` but still it's not the same error message probably.

Reply:
Yes, my bad, you need to assign the exception
```
for _ in range(OCC_MAX_ATTEMPTS):
    cursor = connection.cursor()
    try:
        cursor.execute(query)
        connection.commit()
        break
    except pymonetdb.exceptions.IntegrityError as exc:
        integrity_exc = exc
        connection.rollback()
        continue
    except Exception as exc:
        connection.rollback()
        raise exc
else:
    raise integrity_exc
```
Readability wise it is much better. Just by looking at the overall structure you see than we iterate a number of times and if something doesn't succeed we raise an error. You don't need to look inside the except blocks to realise that some counter is incremented etc. That's exactly why `for` loops in python can have an `else` block. Once you know it it becomes obvious.

Reply:
Yes you are right about the for loop, sorry my mistake.

My comments were about the `for/else`. How will you propagate the same error as the one you received?

Reply:
@ThanKarab Did you read what I wrote the second time?

Reply:
Sorry, I missed the added line.

I would prefer this one but the difference is extremely minor at this point.
```
        for attempt in range(OCC_MAX_ATTEMPTS):
            cursor = connection.cursor()
            try:
                cursor.execute(query)
                connection.commit()
                break
            except pymonetdb.exceptions.IntegrityError as integrity_error:
                connection.rollback()
                # If there are no more attempts
                if attempt == OCC_MAX_ATTEMPTS-1:
                    raise integrity_error
            except Exception as exc:
                connection.rollback()
                raise exc
```

@KFilippopolitis Which one is the prettiest of them all?

Reply:
In this case i think for/else is a little cleaner imo.
```
for _ in range(OCC_MAX_ATTEMPTS):
    cursor = connection.cursor()
    try:
        cursor.execute(query)
        connection.commit()
        break
    except pymonetdb.exceptions.IntegrityError as exc:
        integrity_error = exc
        connection.rollback()
      -----> continue
    except Exception as exc:
        connection.rollback()
        raise exc
else:
    raise integrity_error
```
@jassak Is the continue needed there?